Add magic check for base85

benediktwerner opened the pull request:
I also fixed the `alphabetName` name function since it currently doesn't work. (Can be checked by trying to use `FromBase85` on `z` which should produce 4 zeros if the encoding is `Standard` but currently doesn't because the alphabet/encoding name will be `undefined`.)

I also made `FromBase85` ignore whitespace since Ascii85 is supposed to ignore all whitespace according to [Wikipedia](https://en.wikipedia.org/wiki/Ascii85) (`History > Adobe version`).

I think usually this only applies to the Adobe version but I don't think there is a good reason to not always do it.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/gchq/CyberChef?pullRequest=1033) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/gchq/CyberChef?pullRequest=1033) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/gchq/CyberChef?pullRequest=1033) it.</sub>

n1073645 commented on src/core/operations/FromBase85.mjs:
```diff
@@ -33,6 +33,23 @@ class FromBase85 extends Operation {
         ];
+        this.checks = [
+            {
+                pattern: "^\\s*(?:<~)?(?:(?:\\s*[!-u]){5}|\\s*z)+[!-u\\s]*(?:~>)?\\s*$",
+                flags: "i",
+                args: ["!-u"]
```
Hello, thank you very much for contributing to the magic operation. 

The issue with base85 is that the alphabet spans such a large range, !-u, which includes all the upper case letters, most of the lower case letters and a bunch of punctuation. Thankfully, what it does not include is spaces, we need to modify these regexes so that they only match on inputs which are a certain length. The length should probably be in the region of 15-20, longer than most words. So this regex might become something like:

`^(<~)?(?:([!-u]|z)){20}[!-u]*(~>)?$`

benediktwerner replied:
Well, the Ascii85 i.e. the Adobe version which seems to be the most common allows any whitespace anywhere and it should get ignored.

But I guess it makes sense to avoid spaces in the detection and search for larger continuous blocks. But then it should probably not use `^` and `$`. I guess it also makes sense to ignore the rule that `z` can only appear between groups of 5 other letters. I also just realized the `i` flag doesn't really make sense.

Would something like this be ok? `(?:<~)?\s*(?:[!-uz]){20,}\s*(?:~>)?`

benediktwerner replied:
Actually, thinking about it there is no reason to include the `<~` since it won't change the result and I don't think it will improve performance. Just this should be fine: `(?:[!-uz]){20}`

benediktwerner replied:
Or maybe a better option would be to require at least one block of 20 characters but still allow whitespace everywhere else and enforce that the whole input is valid: `^\s*(?:<~)?[\s!-uz]*[!-uz]{20}[\s!-uz]*(?:~>)?\s*$`.

n1073645 replied:
I'd say it is best to include the delimiter (<~) just to add to the accuracy of the regex. 

Including a space in an example base85 encoded string breaks the fromBase85 operation, which means it would then not be seen as successful by magic. So without further modification to the fromBase85 operation then we shouldn't include it in the regex.

benediktwerner replied:
> I also made FromBase85 ignore whitespace since Ascii85 is supposed to ignore all whitespace according to Wikipedia (History > Adobe version).

I added whitespace support in this PR already.

Including `<~` makes sense but in a regex like this `(?:<~)?\s*[!-uz]{20,}\s*(?:~>)?` it will not make a difference. Since it is optional it only matters when there is something before or after it, like the `^`.

I think this (`^\s*(?:<~)?[\s!-uz]*[!-uz]{20}[\s!-uz]*(?:~>)?\s*$`) should be fine. It will reject everything that doesn't have at least 20 characters of continuous base85 but it still allows arbitrary whitespace otherwise and it also enforces that the whole input doesn't contain any non-base85 characters. It might still match some long uppercase sequences but I think that's unavoidable and I think most base85 strings will at least have some continuous parts or more likely not whitespace at all.

benediktwerner commented:
I now updated the regexes to something more appropriate.

I also changed the base85 operation to only remove the start and end markers (`<~ ... ~>`) in Standard/Ascii85 encoding because I realized they are actually valid and normal characters in the third encoding. They can/should only used with Ascii85.

@n1073645 Any comments on this? Do you think the regexes are now okay?

n1073645 commented:
How I went about writing the regexes for the fromBase operations was, I took the from Base64 regex:

`^\\s*(?:[A-Z\\d+/]{4})+(?:[A-Z\\d+/]{2}==|[A-Z\\d+/]{3}=)?\\s*$`

Then I just modified the alphabet in them/structure if necessary. 

The regex you have proposed looks good, I like the comments explaining each part.

bennofs commented:
The requirement for 20 characters feels quite high to me, also compared to the other encodings which mostly require single-digit amounts. And base85 is high density.
